File: Elixir/binary_search.ex
defmodule BinarySearch do
  def binary_search(lst, target), do: binary_search(lst, target, 0, length(lst))

  def binary_search(_, _, min, stop) when min > stop, do: -1

  def binary_search(lst, target, min, stop) do
      middle = (min + stop) |> div(2) # Breaks list on the middle
      curr_value = Enum.at(lst, middle)

      cond do
        curr_value == target -> middle # found! Return the index
        curr_value > target -> binary_search(lst, target, min, middle - 1) # Runs the left fragment portion of the list
        curr_value < target -> binary_search(lst, target, middle + 1, stop) # Runs the right fragment portion of the list
      end
    end
  end

  BinarySearch.binary_search([1, 2, 3, 4, 5, 6, 7], 10)

File: C#/Binary_search.cs
using System;

class Program
{
    
    static int BinarySearch(int[] arr, int target)
    {
        int left = 0;
        int right = arr.Length - 1;

        while (left <= right)
        {
            int mid = left + (right - left) / 2;

            
            if (arr[mid] == target)
                return mid;

            
            if (arr[mid] > target)
                right = mid - 1;

            
            else
                left = mid + 1;
        }

        return -1;
    }

    static void Main(string[] args)
    {
        int[] sortedArray = { 2, 4, 6, 8, 10, 12, 14, 16, 18, 20 };
        int target = 12;

        int result = BinarySearch(sortedArray, target);

        if (result != -1)
            Console.WriteLine($"Element {target} found at index {result}");
        else
            Console.WriteLine($"Element {target} not found in the array");
    }
}
